Source/CBLRestReplicator+Backgrounding.m

```diff
@@ -65,39 +65,46 @@ - (void) endBackgrounding {
 
 // Called when the replicator goes idle (from -updateActive)
 - (void) okToEndBackgrounding {
-    dispatch_async(dispatch_get_main_queue(), ^{
-        if ([_bgMonitor hasBackgroundTask]) {
-            _deepBackground = YES;
-            [self updateSuspended];
-            LogTo(Sync, @"%@: Now idle; ending background task", self);
-            [_bgMonitor endBackgroundTask];  // will probably suspend the process immediately
-        }
-    });
+    if ([_bgMonitor hasBackgroundTask]) {
+        _deepBackground = YES;
+        [self updateSuspended];
+        LogTo(Sync, @"%@: Now idle; ending background task", self);
+        [_bgMonitor endBackgroundTask];  // will probably suspend the process immediately
+    }
+}
+
+- (void) updateSuspended {
+    BOOL suspended = (_filesystemUnavailable || _deepBackground);
+    self.suspended = suspended;
 }
 
 
 ////// All the methods below are called on the MAIN THREAD, not the replicator thread ////////
 
 
 - (void) appBackgrounding {
-    if (_active && [_bgMonitor beginBackgroundTaskNamed: self.description]) {
-        LogTo(Sync, @"%@: App backgrounding; starting temporary background task", self);
-    } else {
-        LogTo(Sync, @"%@: App backgrounding, but replication is inactive; suspending", self);
-        _deepBackground = YES;
-        [self updateSuspended];
-    }
+    [self.db doSync: ^{
+        if (_active && [_bgMonitor beginBackgroundTaskNamed: self.description]) {
+            LogTo(Sync, @"%@: App backgrounding; starting temporary background task", self);
+        } else {
+            LogTo(Sync, @"%@: App backgrounding, but replication is inactive; suspending", self);
+            _deepBackground = YES;
+            [self updateSuspended];
+        }
+    }];
 }
 
 
 - (void) appForegrounding {
-    BOOL ended = [_bgMonitor endBackgroundTask];
-    if (ended)
-        LogTo(Sync, @"%@: App foregrounded, ending background task", self);
-    if (_deepBackground) {
-        _deepBackground = NO;
-        [self updateSuspended];
-    }
+    [self.db doSync: ^{
+        BOOL ended = [_bgMonitor endBackgroundTask];
```

Comment:
-endBackgroundTask also says "should be called on the main thread."

Reply:
I will revert the change and fix only race condition on accessing the _access var. The rest of the fix is not necessary and not correct per your comment.